### DISCUSSION TITLE
Best method for modifying global shared object safely

### DISCUSSION THREAD

**Discussion by watkinsmike** [fastapi/fastapi#9110] [Questions] [answered]
### Description

I have a global client object that stores credentials that expire at an interval. This object will be used to make an external API call as part of the logic of some requests. I would like to modify this object in a safe manner on an interval so that it can be used safely elsewhere.

### Possible solution
```Python
client = Client(..)

def update_client():
    global client
    client = Client(...)   # updated credentials will be passed to Client

async def loop_forever():
    while True:
        # Would I need locking in this situation?
        # async with asyncio.Lock():  
             # update_client()
        # await asyncio.sleep(3600)
        update_client()
        await asyncio.sleep(3600)

loop = asyncio.get_event_loop()
loop.create_task(loop_forever())

@router.post("/route")
def route(request: Request) -> Response:
    client.client_method()
    ....
```

* Is this a safe and effective to update the client and not cause any race conditions?
* Will adding this forever loop task to the event loop cause any issues with FastAPI operation itself?

### Environment
fastapi==0.61.0
uvicorn==0.11.8
gunicorn==20.0.4

**Comment by YuriiMotov** [comment, marked as answer]
It really depends on the implementation of `Client`.
I would suggest using distributed key-value storage like Redis. Probably in combination with Redis Lock if you need to update several objects in one transaction
